pandas/core/frame.py

```diff
@@ -3892,6 +3893,22 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
         if not isinstance(keys, list):
             keys = [keys]
 
+        missing = []
+        for x in keys:
+            if not (is_scalar(x) or isinstance(x, tuple)):
```

Comment:
why this overly complicated condition? can u show a simple example which fails now?

Reply:
This came out of the 2./3. outlined in #22484 and then the feedback from @gfyoung further up the thread